blockstack/cli-blockstack
# Update to Stacks 2.0 transactions

This PR updates the `send_tokens` and `make_keychain` commands for Stacks 2.0 and adds `deploy_contract` and `call_contract_func` commands to publish and call smart contracts. The PR also adds the bin name `blockstack` so commands can be called using just `blockstack` instead of `blockstack-cli`

`send_tokens` new usage:
```
send_tokens ADDRESS AMOUNT FEE_RATE NONCE PAYMENT_KEY [MEMO] 

$ blockstack send_tokens ST1WZ69T99RHQMQX3D91ZH2R37GV5NK8KDS5D5VDZ 1000 1 0 381314da39a45f43f45ffd33b5d8767d1a38db0da71fea50ed9508e048765cf301 -t
```

To estimate the transaction fee needed without broadcasting the transaction, add the `-e` flag.

To see the raw signed transaction without broadcasting, add the `-x` flag.

`make_keychain` returns
```
{
  "mnemonic": "private unhappy random runway boil scissors remove harvest fatigue inherit inquiry still before mountain pet tail mad accuse second milk client rebuild salt chase",
  "keyInfo": {
    "privateKey": "381314da39a45f43f45ffd33b5d8767d1a38db0da71fea50ed9508e048765cf301",
    "address": "ST1BG7MHW2R524WMF7X8PGG3V45ZN040EB9EW0GQJ",
    "btcAddress": "n4X37UmRZYk9HawtS1w4xRtqJWhByxiz3c",
    "index": 0
  }
}
```

`deploy_contract` usage:
```
deploy_contract SOURCE_FILE CONTRACT_NAME FEE NONCE PAYMENT_KEY

$ blockstack deploy_contract ./my_contract_source.clar hello_world 660 0 381314da39a45f43f45ffd33b5d8767d1a38db0da71fea50ed9508e048765cf301 -t
```
To estimate the transaction fee needed without broadcasting the transaction, add the `-e` flag.
To see the raw signed transaction without broadcasting, add the `-x` flag.

`call_contract_func` usage:

```
call_contract_func CONTRACT_ADDRESS CONTRACT_NAME FUNCTION_NAME FEE NONCE PAYMENT_KEY

blockstack call_contract_func ST1WZ69T99RHQMQX3D91ZH2R37GV5NK8KDS5D5VDZ hello_world my_func 300 0 381314da39a45f43f45ffd33b5d8767d1a38db0da71fea50ed9508e048765cf301 -t
```
To estimate the transaction fee needed without broadcasting the transaction, add the `-e` flag.
To see the raw signed transaction without broadcasting, add the `-x` flag.

The command will fetch the contract ABI and prompt for arguments to the function call. 
```
? Enter value for function argument "myBoolArg" of type bool (Use arrow keys)
❯ True 
  False 
```

## Testing Instructions:
1. Use the `make_keychain` command with the testnet option to generate a new key.
```
blockstack make_keychain -t
```
2. Spin up a local testnet with the side car. Plug the address generated above into the genesis configuration. Alternatively, use the [testnet faucet](https://testnet.blockstack.org/faucet) to send tokens to the address generated above. 

3. Use the `send_tokens` command to send tokens to any address. Using the `-t` option for testnet. By default using the `-t` flag sets the Stacks blockchain node that the CLI communicates with to `http://neon.blockstack.org:20443`. You can specify an alternate node using: `-H "<Your local core node URL>"`

```
blockstack send_tokens ST16KX11KMXWG9NRVCM1D5CHZGR3HFRQBK0DN9XAT 10000 300 0 381314da39a45f43f45ffd33b5d8767d1a38db0da71fea50ed9508e048765cf301 -t 
```

4. Verify balances using the `balance` command
```
blockstack balance ST16KX11KMXWG9NRVCM1D5CHZGR3HFRQBK0DN9XAT -t
```
You can also specify a node to look up balances on using the `-H` flag here.

5. Test contract deploy using the `deploy_contract` command:
```
blockstack deploy_contract ./my_contract_source.clar hello_world 800 1 381314da39a45f43f45ffd33b5d8767d1a38db0da71fea50ed9508e048765cf301 -t
```

6. Test calling the function that was just deployed using the `call_contract_func` command:
```
blockstack call_contract_func ST16KX11KMXWG9NRVCM1D5CHZGR3HFRQBK0DN9XAT hello_world test_function 400 2 381314da39a45f43f45ffd33b5d8767d1a38db0da71fea50ed9508e048765cf301 -t
```
Verify that function was successfully called and processed.
